schleuder: init

mweinelt opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Providing a gpg encrypted remailer using https://schleuder.org/.

Three parts and I need help with ruby packaging.
- [x] https://0xacab.org/schleuder/schleuder
  - was simple, could be pulled from rubygems.org
- [x] https://0xacab.org/schleuder/schleuder-cli
  - only available from Git and I fail to provide the correct Gemfile, so that the `schleuder-cli-0.1.0` gem does not appear in the gemset
    ```
    find: '/nix/store/hw8s9qi6kridfkif10hwk66w185acgcz-schleuder-cli-0.1.0/lib/ruby/gems/2.7.0/gems/schleuder-cli-0.1.0': No such file or directory
    ```
- [ ] https://0xacab.org/schleuder/schleuder-web (TODO)

When that is done:
- [x] NixOS module (WIP)
- [x] NixOS test (WIP)

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

mweinelt committed:
schleuder: init at 0.4.2

mweinelt committed:
schleuder-cli: init 0.1.0

Mic92 commented:
```
diff --git a/pkgs/tools/security/schleuder-cli/default.nix b/pkgs/tools/security/schleuder-cli/default.nix
index d05d4f841df..e0ee2baa657 100644
--- a/pkgs/tools/security/schleuder-cli/default.nix
+++ b/pkgs/tools/security/schleuder-cli/default.nix
@@ -15,6 +15,8 @@ bundlerApp {
     "schleuder-cli"
   ];
 
+  installManpages = false;
+
   passthru.updateScript = bundlerUpdateScript "schleuder-cli";
 
   meta = with lib; {
```

lheckemann committed:
nixos/schleuder: init module and accompanying test

Co-Authored-By: Martin Weinelt <hexa@darmstadt.ccc.de>
Co-Authored-By: Cole Helbling <cole.helbling@determinate.systems>

mweinelt commented on nixos/modules/services/mail/schleuder.nix:
```diff
@@ -0,0 +1,100 @@
+      };
+    */
+    settings = lib.mkOption {
+      description = "Settings for schleuder.yml";
```
```suggestion
      description = ''
        Your <filename>schleuder.yml</filename> as a Nix attribute set.
        
        Check the <link xlink:href="https://0xacab.org/schleuder/schleuder/blob/master/etc/schleuder.yml">example configuration</link> for possible values.
      ''
```

Users love clickable links!

mweinelt commented on nixos/modules/services/mail/schleuder.nix:
```diff
@@ -0,0 +1,100 @@
+      # We would have liked to use DynamicUser, but since the default
+      # database is SQLite and lives in StateDirectory, and that same
+      # database needs to be readable from the postfix service, this
+      # isn't trivial to do.
```
What I see from the `master.conf` is that it tries to execute `schleuder` as the `schleuder` user, which is something the `master` process can reasonably do, since it runs as root. Not sure where the blocker here is.

lheckemann replied:
That the schleuder user doesn't exist, and postfix doesn't know how to set up the `/var/lib/private/schleuder -> /var/lib/schleuder` bind mount that's necessary for the user to access schleuder's StateDirectory.

lheckemann replied:
Oh right, that's on the assumption that the user isn't created when using DynamicUser. The bind mount is still a problem regardless.

mweinelt commented on nixos/modules/services/mail/schleuder.nix:
```diff
@@ -0,0 +1,100 @@
+      default = { };
+    };
+    listDefaults = lib.mkOption {
+      description = "Default settings for lists (list-defaults.yml)";
```
```suggestion
      description = ''
        Your <filename>lists-default.yml</filename> as a Nix attribute set.
        
        Check the <link xlink:href="https://0xacab.org/schleuder/schleuder/-/blob/master/etc/list-defaults.yml">example configuration</link> for possible values.
      '';
```

mweinelt commented on nixos/modules/services/mail/schleuder.nix:
```diff
@@ -0,0 +1,100 @@
+      default = [];
+      example = ["securelists.example.com"];
+      };
+    */
```
Not sure if this would be sane, but could be parsed from the list of lists.

Also where would this need to be configured? Looks to me like schleuder doesn't really care about domains.

lheckemann replied:
See https://schleuder.org/schleuder/docs/server-admins.html#configuration -- dedicating an entire domain to schleuder is suggested. This isn't appropriate for the use case currently tested, where users and secure lists coexist on the same domain.

lheckemann replied:
Made the description of the unimplemented option clearer :)

mweinelt commented on nixos/modules/services/mail/schleuder.nix:
```diff
@@ -0,0 +1,100 @@
+  options.services.schleuder = {
+    enable = lib.mkEnableOption "Schleuder secure remailer";
+    lists = lib.mkOption {
+      description = "Addresses of lists that should be handled by Schleuder.";
```
```suggestion
      description = "List of list addresses that should be handled by Schleuder.";
```

This is otherwise easily misunderstood.

mweinelt commented on nixos/modules/services/mail/schleuder.nix:
```diff
@@ -0,0 +1,100 @@
+    */
+    settings = lib.mkOption {
+      description = "Settings for schleuder.yml";
+      type = settingsFormat.type;
+      default = { };
```
Probably best to go for a submodule with a `freeformType = format.type` instead, so we can have some sane presets.

I wouldn't want anyone to be inspired by https://0xacab.org/schleuder/schleuder/-/blob/master/etc/schleuder.yml#L34. We can do such much better.

mweinelt replied:
That preset wouldn't be visible on search.nixos.org.

mweinelt reviewed (commented):
Awesome work, consider this a first pass and I'll gladly work on some of the issues myself.

mweinelt commented:
Dropped the `schleuder-web` package, which was not useful at all and broke eval.

lheckemann commented on nixos/tests/schleuder.nix:
```diff
@@ -0,0 +1,96 @@
+        sudo -u alice gpg --armor --export alice@${domain} > alice.asc
+        schleuder-cli list new security@${domain} alice@${domain} alice.asc
+
+        msmtp --auto-from security@${domain} --host=${domain} --port=25 --tls --tls-starttls <<EOF
+        Subject: really big security issue!!
+        From: person@${domain}
+
+        I found a big security problem!
+        EOF
```
Why does this work? I'd expect it not to o.o

mweinelt replied:
I would expect that to work, given that it is wrapped in EOF. I didn't test it.

lheckemann replied:
Bash doesn't strip the leading space, but somehow msmtp seems to be fine with it. Bash would strip leading tabs with `<<-EOF` but, well, tabs.

mweinelt replied:
I'd expect mail to be fine with it.

lheckemann commented on nixos/tests/schleuder.nix:
```diff
@@ -0,0 +1,96 @@
+        sudo -u alice gpg --armor --export alice@${domain} > alice.asc
+        schleuder-cli list new security@${domain} alice@${domain} alice.asc
+
+        msmtp --auto-from security@${domain} --host=${domain} --port=25 --tls --tls-starttls <<EOF
```
It deduces the envelope-from from the header-from, in this case "person@acme.test". It's fairly irrelevant, I'm just used to using msmtp with `--auto-from`, we can take this out if you like.

mweinelt replied:
That or add a comment.

lheckemann commented on nixos/tests/schleuder.nix:
```diff
@@ -0,0 +1,96 @@
+
+        I found a big security problem!
+        EOF
+        sleep 5
+        mail=$(find /var/spool/mail -type f)
```
Yes. The email flows as follows:
```
msmtp
  ⬇️
postfix
  ⬇️
schleuder (resends it encrypted)
  ⬇️
postfix
  ⬇️
maildir
```
So really, we want to wait until it's in the maildir. I'll implement that more cleanly than with a sleep :)

lheckemann commented on nixos/tests/schleuder.nix:
```diff
@@ -0,0 +1,96 @@
+      settings.api = {
+        tls_cert_file = "${certs.${domain}.cert}";
+        tls_key_file = "${certs.${domain}.key}";
+        valid_api_keys = [ "fnord" ];
```
I've added assertions for the obvious secrets _not_ to be defined in NixOS config, as well as implementing and using the `extraSettingsFile` mechanism in the test.

lheckemann committed:
bundler bin stubs: Squelch sudo warning

@ruby maintainers: please feel free to revert this and let me know,
should it cause any problems.

Stuff that's in the store shouldn't be writable! So let's disable
Bundler's requires_sudo? method, which checks if some files are
writable and suggests using sudo if not, entirely.

Previously, schleuder-cli would print to stderr when run:

Following files may not be writable, so sudo is needed:
  /nix/store/qg40x0ysrf9x6sag6qgb1klg87lskdp5-schleuder-cli-0.1.0/lib/ruby/gems/2.7.0
  /nix/store/qg40x0ysrf9x6sag6qgb1klg87lskdp5-schleuder-cli-0.1.0/lib/ruby/gems/2.7.0/bin
  /nix/store/qg40x0ysrf9x6sag6qgb1klg87lskdp5-schleuder-cli-0.1.0/lib/ruby/gems/2.7.0/bin
  /nix/store/qg40x0ysrf9x6sag6qgb1klg87lskdp5-schleuder-cli-0.1.0/lib/ruby/gems/2.7.0/build_info
  /nix/store/qg40x0ysrf9x6sag6qgb1klg87lskdp5-schleuder-cli-0.1.0/lib/ruby/gems/2.7.0/bundler
  /nix/store/qg40x0ysrf9x6sag6qgb1klg87lskdp5-schleuder-cli-0.1.0/lib/ruby/gems/2.7.0/cache
  /nix/store/qg40x0ysrf9x6sag6qgb1klg87lskdp5-schleuder-cli-0.1.0/lib/ruby/gems/2.7.0/doc
  /nix/store/qg40x0ysrf9x6sag6qgb1klg87lskdp5-schleuder-cli-0.1.0/lib/ruby/gems/2.7.0/extensions
  /nix/store/qg40x0ysrf9x6sag6qgb1klg87lskdp5-schleuder-cli-0.1.0/lib/ruby/gems/2.7.0/gems
  /nix/store/qg40x0ysrf9x6sag6qgb1klg87lskdp5-schleuder-cli-0.1.0/lib/ruby/gems/2.7.0/plugins
  /nix/store/qg40x0ysrf9x6sag6qgb1klg87lskdp5-schleuder-cli-0.1.0/lib/ruby/gems/2.7.0/specifications

mweinelt commented on nixos/modules/services/mail/schleuder.nix:
```diff
@@ -0,0 +1,162 @@
+      < ${settingsFormat.generate "schleuder.yml" cfg.settings} \
+      '. * $overrides[0]' \
+      > /etc/schleuder/schleuder.yml
+    chown schleuder /etc/schleuder/schleuder.yml
```
```suggestion
    chown schleuder: /etc/schleuder/schleuder.yml
```

Only user? Also group?

lheckemann replied:
Done

mweinelt reviewed (commented):
Missing release notes entry and tests were not added to `all-tests.nix`

mweinelt commented:
And kindly split the test into a separate commit and in that commit set up passthru.tests in the packages tested.

lheckemann committed:
rl-2211: document schleuder addition
